posthog/queries/retention.py

```diff
@@ -1,62 +1,203 @@
+import random
+import string
 from datetime import timedelta
 from typing import Any, Dict, List, Tuple, Union
 
 from dateutil.relativedelta import relativedelta
+from django.core.cache import cache
+from django.db import connection
+from django.db.models.functions.datetime import TruncDay, TruncHour, TruncMonth, TruncWeek
+from django.db.models.query import QuerySet
 from django.utils.timezone import now
 
+from posthog.constants import TREND_FILTER_TYPE_ACTIONS, TREND_FILTER_TYPE_EVENTS
 from posthog.models import Event, Filter, Team
+from posthog.models.entity import Entity
+from posthog.models.utils import namedtuplefetchall
 from posthog.queries.base import BaseQuery
+from posthog.utils import generate_cache_key
 
 
 class Retention(BaseQuery):
-    def calculate_retention(self, filter: Filter, team: Team, total_intervals=11):
-        def _determineTimedelta(
-            total_intervals: int, period: str
-        ) -> Tuple[Union[timedelta, relativedelta], Union[timedelta, relativedelta]]:
-            if period == "Hour":
-                return timedelta(hours=total_intervals), timedelta(hours=1)
-            elif period == "Week":
-                return timedelta(weeks=total_intervals), timedelta(weeks=1)
-            elif period == "Month":
-                return relativedelta(months=total_intervals), relativedelta(months=1)
-            elif period == "Day":
-                return timedelta(days=total_intervals), timedelta(days=1)
-            else:
-                raise ValueError(f"Period {period} is unsupported.")
-
+    def preprocess_params(self, filter: Filter, total_intervals=11):
         period = filter.period or "Day"
-
-        tdelta, t1 = _determineTimedelta(total_intervals, period)
-
+        tdelta, t1 = self.determineTimedelta(total_intervals, period)
         filter._date_to = ((filter.date_to if filter.date_to else now()) + t1).isoformat()
 
         if period == "Hour":
             date_to = filter.date_to if filter.date_to else now()
+            date_from: datetime.datetime = date_to - tdelta
+        elif period == "Week":
+            date_to = (filter.date_to if filter.date_to else now()).replace(hour=0, minute=0, second=0, microsecond=0)
             date_from = date_to - tdelta
+            date_from = date_from - timedelta(days=date_from.isoweekday() % 7)
         else:
-
             date_to = (filter.date_to if filter.date_to else now()).replace(hour=0, minute=0, second=0, microsecond=0)
             date_from = date_to - tdelta
 
         filter._date_from = date_from.isoformat()
         filter._date_to = date_to.isoformat()
 
-        resultset = Event.objects.query_retention(filter, team)
+        entity = (
+            Entity({"id": "$pageview", "type": TREND_FILTER_TYPE_EVENTS})
+            if not filter.target_entity
+            else filter.target_entity
+        )
+
+        # need explicit handling of date_from so it's not optional but also need filter object for date_filter_Q
```

Comment:
I'm going to either fix the typing or refactor the filter object a slight bit because the optional typing of date_from becomes unwieldy on the filter object. We usually end up setting filter._date_from but then when calling anything along the lines of filter.date_from.strftime we get hit with the optional type issue